RTC: STM32: Add alarm support for all STM32 MCU series

ajarmouni-st opened the pull request:
This PR implements `rtc_driver_api` Alarm functions for the 2 alarms, A & B, of STM32 RTC device.
Tested on all the boards that have been added in the `tests/drivers/rtc/rtc_api/boards`.

Edit: STM32F1x SoC series is not supported.

ajarmouni-st committed:
dts: bindings: rtc: stm32-rtc: add alarms-count property

Add alarms-count property to STM32 RTC binding

Signed-off-by: Abderrahmane Jarmouni <abderrahmane.jarmouni-ext@st.com>

ajarmouni-st committed:
dts: arm: st: stm32-rtc: add alarms-count property

Add alarms-count to STM32 RTC node of all series except F1X.

Signed-off-by: Abderrahmane Jarmouni <abderrahmane.jarmouni-ext@st.com>

ajarmouni-st committed:
boards: st: configure STM32 RTC

Configure STM32 RTC for a large selction of boards.

Signed-off-by: Abderrahmane Jarmouni <abderrahmane.jarmouni-ext@st.com>

ajarmouni-st committed:
tests: drivers: rtc: rtc_api: support STM32 boards

Add RTC config for a large selection of STM32 boards.

Signed-off-by: Abderrahmane Jarmouni <abderrahmane.jarmouni-ext@st.com>

ajarmouni-st committed:
tests: drivers: rtc_api: test_alarm: add alarm id in logs

Add alarm ID in debug & error logs.

Signed-off-by: Abderrahmane Jarmouni <abderrahmane.jarmouni-ext@st.com>

decsny commented on dts/bindings/rtc/st,stm32-rtc.yaml:
```diff
@@ -18,3 +21,9 @@ properties:
+    description: |
+      Number of alarms supported by STM32 RTC device.
+      Most of STM32 MCU series have 2 RTC alarms, A & B.
+      Defaults to 0 when not declared in devicetree.
```
either put a default: value here or make it required: true

ajarmouni-st replied:
There is already a default value in `rtc-device.yaml`.

decsny replied:
ok

decsny commented:
why are there two PRs ?  #71965

gautierg-st commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -76,46 +131,179 @@ struct rtc_stm32_config {
+	if (id == RTC_STM32_ALRM_A) {
+		LL_RTC_EnableIT_ALRA(rtc);
+	} else {
+		LL_RTC_EnableIT_ALRB(rtc);
```
Did you try to compile the alarm subsystem on series with only one alarm other than F1 (like C0 or F0) ? I don't think it works since all functions and variables with ALRB aren't defined.

ajarmouni-st replied:
There is a `CONFIG_SOC_SERIES_STM32F1X` just before.

ajarmouni-st replied:
Good catch !!

ajarmouni-st replied:
Fixed.

gautierg-st commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -322,6 +576,482 @@ static int rtc_stm32_get_time(const struct device *dev, struct rtc_time *timeptr
+	return err;
+}
+
+static int rtc_stm32_alarm_set_time(const struct device *dev, uint16_t id, uint16_t mask,
```
This function is very long, is there a way to split it in smaller pieces?

ajarmouni-st replied:
Fixed

ajarmouni-st commented:
> why are there two PRs ? #71965

A pure coincidence :)

gautierg-st reviewed (changes requested):
I haven't reviewed everything but I have already a few comments

bjarki-trackunit commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -322,6 +576,482 @@ static int rtc_stm32_get_time(const struct device *dev, struct rtc_time *timeptr
+		ll_mask = LL_RTC_ALMB_MASK_ALL;
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_SECOND) {
+		if (timeptr->tm_sec > 59) {
```
Have you seen 
https://github.com/zephyrproject-rtos/zephyr/blob/35429debb627943410f3c3651f267fb40f04a9cf/drivers/rtc/rtc_utils.c#L15C6-L15C33 ? unless the alarms have some additional constraint on the time fields, it can save a bit of repeated code :)

bjarki-trackunit commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -322,6 +576,482 @@ static int rtc_stm32_get_time(const struct device *dev, struct rtc_time *timeptr
+		goto unlock;
+	}
+
+	if (p_rtc_alrm->user_callback != NULL) {
+		LOG_ERR("Alarm %d is busy", id);
+		err = -EBUSY;
+		goto unlock;
+	}
```
It is allowed to overwrite an existing alarm, either to disable or reconfigure it :) It is not explicitly mentioned in the documentation, maybe it should be?

ajarmouni-st replied:
Fixed

bjarki-trackunit commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -67,7 +75,54 @@ LOG_MODULE_REGISTER(rtc_stm32, CONFIG_RTC_LOG_LEVEL);
+
 struct rtc_stm32_config {
+#ifdef CONFIG_RTC_ALARM
+	uint32_t alarms_count;
```
A bit overkill with a u32 :) maybe uint8 or uint16 (to match with the size of id in `rtc_alarm_set_time`)

ajarmouni-st replied:
Replaced by `RTC_STM32_ALARMS_COUNT` in `rtc_ll_stm32.h`

erwango commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -322,6 +576,482 @@ static int rtc_stm32_get_time(const struct device *dev, struct rtc_time *timeptr
+
+#if defined(CONFIG_SOC_SERIES_STM32H7X) && defined(CONFIG_CPU_CORTEX_M4)
+	LL_C2_EXTI_ClearFlag_0_31(RTC_EXTI_LINE);
+#elif defined(CONFIG_SOC_SERIES_STM32C0X) \
+	|| defined(CONFIG_SOC_SERIES_STM32G0X) \
+	|| defined(CONFIG_SOC_SERIES_STM32L5X) \
+	|| defined(CONFIG_SOC_SERIES_STM32H5X)
```
#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32g0_exti)

ajarmouni-st replied:
Fixed

erwango commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -67,7 +75,54 @@ LOG_MODULE_REGISTER(rtc_stm32, CONFIG_RTC_LOG_LEVEL);
+#define RTC_STM32_ALRM_A	0U
+#define RTC_STM32_ALRM_B	1U
+
+#ifdef CONFIG_SOC_SERIES_STM32L4X
+#define RTC_EXTI_LINE	LL_EXTI_LINE_18
+#elif defined(CONFIG_SOC_SERIES_STM32C0X) \
+	|| defined(CONFIG_SOC_SERIES_STM32G0X)
+#define RTC_EXTI_LINE	LL_EXTI_LINE_19
+#elif defined(CONFIG_SOC_SERIES_STM32F4X) \
+	|| defined(CONFIG_SOC_SERIES_STM32F0X) \
+	|| defined(CONFIG_SOC_SERIES_STM32F1X) \
+	|| defined(CONFIG_SOC_SERIES_STM32F2X) \
+	|| defined(CONFIG_SOC_SERIES_STM32F3X) \
+	|| defined(CONFIG_SOC_SERIES_STM32F7X) \
+	|| defined(CONFIG_SOC_SERIES_STM32WBX) \
+	|| defined(CONFIG_SOC_SERIES_STM32G4X) \
+	|| defined(CONFIG_SOC_SERIES_STM32L0X) \
+	|| defined(CONFIG_SOC_SERIES_STM32L1X) \
+	|| defined(CONFIG_SOC_SERIES_STM32L5X) \
+	|| defined(CONFIG_SOC_SERIES_STM32H7X) \
+	|| defined(CONFIG_SOC_SERIES_STM32H5X) \
+	|| defined(CONFIG_SOC_SERIES_STM32WLX)
+#define RTC_EXTI_LINE	LL_EXTI_LINE_17
+#endif /* CONFIG_SOC_SERIES_STM32L4X */
+#endif /* CONFIG_RTC_ALARM */
```
Please use device tree to specify line number

ajarmouni-st replied:
Done.

erwango commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -67,7 +75,54 @@ LOG_MODULE_REGISTER(rtc_stm32, CONFIG_RTC_LOG_LEVEL);
+#endif /* CONFIG_SOC_SERIES_STM32L4X */
+#endif /* CONFIG_RTC_ALARM */
+
+#if DT_INST_CLOCKS_CELL_BY_IDX(0, 1, bus) == STM32_SRC_LSI
+/* LSI */
+#define RTCCLK_FREQ STM32_LSI_FREQ
+#else
+/* LSE */
+#define RTCCLK_FREQ STM32_LSE_FREQ
+#endif /* DT_INST_CLOCKS_CELL_BY_IDX(0, 1, bus) == STM32_SRC_LSI */
```
Would be cleaner to use proper clock_control API

erwango commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -76,46 +131,179 @@ struct rtc_stm32_config {
-	LL_RTC_DisableWriteProtection(RTC);
+static inline uint32_t ll_func_is_active_alarm(RTC_TypeDef *rtc, uint16_t id)
+{
+#ifdef CONFIG_SOC_SERIES_STM32F1X
+	ARG_UNUSED(id);
+
+	return LL_RTC_IsActiveFlag_ALR(rtc);
+#else
+	if (id == RTC_STM32_ALRM_A) {
+		return LL_RTC_IsActiveFlag_ALRA(rtc);
+	}
+	if (id == RTC_STM32_ALRM_B) {
+		return LL_RTC_IsActiveFlag_ALRB(rtc);
+	}
+#endif /* CONFIG_SOC_SERIES_STM32F1X */
```
Can we use a specific header file to abstract these?
Similar to what is done for STM32 spi for instance.

Having those #ifdef CONFIG_SOC_SERIES_STM32F1X all over the file doesn't help in readability.

ajarmouni-st replied:
Done. Dropped support for F1X series in this driver.

erwango commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -76,46 +131,179 @@ struct rtc_stm32_config {
+	ARG_UNUSED(id);
+
+	LL_RTC_DisableIT_ALR(rtc);
+#else
+	if (id == RTC_STM32_ALRM_A) {
+		LL_RTC_DisableIT_ALRA(rtc);
+	} else {
+		LL_RTC_DisableIT_ALRB(rtc);
+	}
+#endif /* CONFIG_SOC_SERIES_STM32F1X */
+}
+
+static inline void ll_func_enable_alarm(RTC_TypeDef *rtc, uint16_t id)
+{
+#ifdef CONFIG_SOC_SERIES_STM32F1X
+	ARG_UNUSED(rtc);
+	ARG_UNUSED(id);
+#else
+	if (id == RTC_STM32_ALRM_A) {
+		LL_RTC_ALMA_Enable(rtc);
+	} else {
+		LL_RTC_ALMB_Enable(rtc);
+	}
+#endif /* CONFIG_SOC_SERIES_STM32F1X */
+}
+
+static inline void ll_func_disable_alarm(RTC_TypeDef *rtc, uint16_t id)
+{
+#ifdef CONFIG_SOC_SERIES_STM32F1X
+	ARG_UNUSED(rtc);
+	ARG_UNUSED(id);
+#else
+	if (id == RTC_STM32_ALRM_A) {
+		LL_RTC_ALMA_Disable(rtc);
+	} else {
+		LL_RTC_ALMB_Disable(rtc);
+	}
+#endif /* CONFIG_SOC_SERIES_STM32F1X */
+}
+#endif /* CONFIG_RTC_ALARM */
```
Good but, please put these in a dedicated header file

ajarmouni-st replied:
Done.

erwango reviewed (changes requested):
Some initial comments

ajarmouni-st committed:
dts: bindings: rtc: stm32-rtc: add alrm-exti-line property

Gives number of the Extended Interrupts and Event Controller (EXTI)
interrupt line connected to the RTC Alarm event.
Used on all series, except WBAX & U5X, where RTC Alarm interrupt is
routed directly to Nested Vectored Interrupt Controller (NVIC) and to
Power Control (PWR) wake-up pins.

Signed-off-by: Abderrahmane Jarmouni <abderrahmane.jarmouni-ext@st.com>

ajarmouni-st committed:
dts: arm: st: stm32-rtc: add alrm-exti-line property

Add alrm-exti-line to STM32 RTC node of concerned series.

Signed-off-by: Abderrahmane Jarmouni <abderrahmane.jarmouni-ext@st.com>

bjarki-trackunit commented:
After #71784 was merged, the y2k test is now skipped automatically if the RTC correctly returns -EINVAL if the time is outside if its range, like before 2000 :) Could you try try to rebase and run the test suite without the commit excluding the y2k test?

ajarmouni-st commented:
> After #71784 was merged, the y2k test is now skipped automatically if the RTC correctly returns -EINVAL if the time is outside if its range, like before 2000 :) Could you try try to rebase and run the test suite without the commit excluding the y2k test?

Done.

erwango commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -67,55 +78,112 @@ LOG_MODULE_REGISTER(rtc_stm32, CONFIG_RTC_LOG_LEVEL);
 	const struct rtc_stm32_config *cfg = dev->config;
-
+	LL_RTC_InitTypeDef ll_rtc_cfg;
+	int32_t hour_format     = LL_RTC_GetHourFormat(RTC);
+	int32_t sync_prescaler  = LL_RTC_GetSynchPrescaler(RTC);
+	int32_t async_prescaler = LL_RTC_GetAsynchPrescaler(RTC);
```
Why changing to `int32_t` ? these API return uint32

erwango commented on drivers/rtc/rtc_ll_stm32.c:
These are too much changes for a single commit, mixing cosmetic and functional changes.
Please split changes into atomic commits.

erwango commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -22,15 +23,25 @@
-/* subsecond counting is not supported by some STM32L1x MCUs */
+#include "rtc_ll_stm32.h"
+
+/* subsecond counting is not supported by some STM32L1x MCUs (Cat.1) */
+#if (defined(CONFIG_SOC_SERIES_STM32L1X) && !defined(RTC_SUBSECOND_SUPPORT)) \
+	|| defined(CONFIG_SOC_SERIES_STM32F2X)
```
Not related to alarm support, please split into a dedicated commit

erwango commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -147,28 +265,44 @@ static int rtc_stm32_init(const struct device *dev)
 
 	k_mutex_init(&data->lock);
 
-	/* Enable Backup access */
-#if defined(PWR_CR_DBP) || defined(PWR_CR1_DBP) || defined(PWR_DBPCR_DBP) || defined(PWR_DBPR_DBP)
+#if RTC_STM32_BACKUP_DOMAIN_WRITE_PROTECTION
 	LL_PWR_EnableBkUpAccess();
-#endif /* PWR_CR_DBP || PWR_CR1_DBP || PWR_DBPR_DBP */
+#endif /* RTC_STM32_BACKUP_DOMAIN_WRITE_PROTECTION */
```
Not directly linked to alarm support, please make it a dedicated commit

ajarmouni-st replied:
Could be made clean in another PR.

erwango replied:
I prefer to get it now to be sure it is done ;)

erwango commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -147,28 +265,44 @@ static int rtc_stm32_init(const struct device *dev)
 
 	k_mutex_init(&data->lock);
 
-	/* Enable Backup access */
-#if defined(PWR_CR_DBP) || defined(PWR_CR1_DBP) || defined(PWR_DBPCR_DBP) || defined(PWR_DBPR_DBP)
+#if RTC_STM32_BACKUP_DOMAIN_WRITE_PROTECTION
 	LL_PWR_EnableBkUpAccess();
-#endif /* PWR_CR_DBP || PWR_CR1_DBP || PWR_DBPR_DBP */
+#endif /* RTC_STM32_BACKUP_DOMAIN_WRITE_PROTECTION */
 
 	/* Enable RTC clock source */
 	if (clock_control_configure(clk, (clock_control_subsys_t)&cfg->pclken[1], NULL) != 0) {
 		LOG_ERR("clock configure failed\n");
 		return -EIO;
 	}
-
+/*
+ * On STM32WBAX series, there is no bit in BCDR register to enable RTC.
+ * Enabling RTC is done directly via the RCC APB register bit.
+ */
+#ifndef CONFIG_SOC_SERIES_STM32WBAX
```
Requires a dedicated commit.

erwango commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -247,22 +382,29 @@ static int rtc_stm32_set_time(const struct device *dev, const struct rtc_time *t
 static int rtc_stm32_get_time(const struct device *dev, struct rtc_time *timeptr)
 {
 	struct rtc_stm32_data *data = dev->data;
-
 	uint32_t rtc_date, rtc_time;
-
 #if HW_SUBSECOND_SUPPORT
 	const struct rtc_stm32_config *cfg = dev->config;
 	uint32_t rtc_subsecond;
-#endif
+#endif	/* HW_SUBSECOND_SUPPORT */
+	int err;
 
-	int err = k_mutex_lock(&data->lock, K_NO_WAIT);
+	if (timeptr == NULL) {
+		LOG_ERR("NULL rtc_time pointer");
+		return -EINVAL;
+	}
+
+	memset(timeptr, -1, sizeof(struct rtc_time));
+
+	err = k_mutex_lock(&data->lock, K_NO_WAIT);
```
Justification ?
Likely requires a dedicated commit as well.

ajarmouni-st replied:
> Justification ?

We cannot operate on a NULL pointer !?

erwango commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -247,22 +382,29 @@ static int rtc_stm32_set_time(const struct device *dev, const struct rtc_time *t
 static int rtc_stm32_get_time(const struct device *dev, struct rtc_time *timeptr)
 {
 	struct rtc_stm32_data *data = dev->data;
-
 	uint32_t rtc_date, rtc_time;
-
 #if HW_SUBSECOND_SUPPORT
 	const struct rtc_stm32_config *cfg = dev->config;
 	uint32_t rtc_subsecond;
-#endif
+#endif	/* HW_SUBSECOND_SUPPORT */
+	int err;
 
-	int err = k_mutex_lock(&data->lock, K_NO_WAIT);
+	if (timeptr == NULL) {
+		LOG_ERR("NULL rtc_time pointer");
+		return -EINVAL;
+	}
+
+	memset(timeptr, -1, sizeof(struct rtc_time));
+
+	err = k_mutex_lock(&data->lock, K_NO_WAIT);
 
 	if (err) {
 		return err;
 	}
 
 	if (!LL_RTC_IsActiveFlag_INITS(RTC)) {
-		/* INITS flag is set when the calendar has been initialiazed. This flag is
+		/*
+		 * INITS flag is set when the calendar has been initialiazed. This flag is
```
Not related to alarm. Dedicated commit required.

erwango commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -313,15 +458,271 @@ static int rtc_stm32_get_time(const struct device *dev, struct rtc_time *timeptr
-#endif
+#endif /* HW_SUBSECOND_SUPPORT */
 
-	/* unknown values */
-	timeptr->tm_yday  = -1;
-	timeptr->tm_isdst = -1;
```
Justification ?

ajarmouni-st replied:
Line 397: `memset(timeptr, -1, sizeof(struct rtc_time));`

erwango commented on drivers/rtc/rtc_ll_stm32.h:
```diff
@@ -0,0 +1,318 @@
+			ll_mask &= ~LL_RTC_ALMA_MASK_DATEWEEKDAY;
+			p_rtc_alarm->AlarmDateWeekDaySel = LL_RTC_ALMA_DATEWEEKDAYSEL_WEEKDAY;
+		}
+#if RTC_STM32_ALARMS_COUNT > 1
+		if (id == RTC_STM32_ALRM_B) {
+			ll_mask &= ~LL_RTC_ALMB_MASK_DATEWEEKDAY;
+			p_rtc_alarm->AlarmDateWeekDaySel = LL_RTC_ALMB_DATEWEEKDAYSEL_WEEKDAY;
+		}
+#endif /* RTC_STM32_ALARMS_COUNT > 1 */
+
+		if (timeptr->tm_wday == 0) {
+			/* sunday (tm_wday = 0) is not represented by the same value in hardware */
+			p_rtc_alarm->AlarmDateWeekDay = LL_RTC_WEEKDAY_SUNDAY;
+		} else {
+			/* all the other values are consistent with what is expected by hardware */
+			p_rtc_alarm->AlarmDateWeekDay = bin2bcd(timeptr->tm_wday);
+		}
+
+	} else if (mask & RTC_ALARM_TIME_MASK_MONTHDAY) {
+		/* the Alarm compares with the day number & ignores the day of the week */
+		if (id == RTC_STM32_ALRM_A) {
+			ll_mask &= ~LL_RTC_ALMA_MASK_DATEWEEKDAY;
+			p_rtc_alarm->AlarmDateWeekDaySel = LL_RTC_ALMA_DATEWEEKDAYSEL_DATE;
+		}
+#if RTC_STM32_ALARMS_COUNT > 1
+		if (id == RTC_STM32_ALRM_B) {
+			ll_mask &= ~LL_RTC_ALMB_MASK_DATEWEEKDAY;
+			p_rtc_alarm->AlarmDateWeekDaySel = LL_RTC_ALMB_DATEWEEKDAYSEL_DATE;
+		}
+#endif /* RTC_STM32_ALARMS_COUNT > 1 */
+
+		p_rtc_alarm->AlarmDateWeekDay = bin2bcd(timeptr->tm_mday);
+	}
+
+	p_rtc_alrm_time->TimeFormat = LL_RTC_TIME_FORMAT_AM_OR_24;
+
+	p_rtc_alarm->AlarmMask = ll_mask;
+
+	return 0;
+}
```
This function is too long to be in a header.

Note that (unfortunately poorly named) ll_func functions used accross previous drivers are reserved to hw abstraction. 
My initial request to use these functions in previous patch set was targeting functions providing abstraction for F1 series. This is not the case anymore, so naming should be reviewed and functions can lie in driver c file when not providing abstraction.
This is not the case here, so you should use proper dedicated function names.

erwango commented on boards/st/nucleo_f722ze/nucleo_f722ze.dts:
```diff
@@ -174,7 +179,7 @@ zephyr_udc0: &usbotg_fs {
 
 &rtc {
 	clocks = <&rcc STM32_CLOCK_BUS_APB1 0x10000000>,
-		<&rcc STM32_SRC_LSI RTC_SEL(2)>;
+		<&rcc STM32_SRC_LSE RTC_SEL(1)>;
```
Why moving to LSE ?

ajarmouni-st replied:
More precise.

erwango replied:
This should be mentioned in the commit message. Also we need to keep at least one test on the LSI (could be an overlay in the test)

erwango reviewed (changes requested):
One additional comment on `tests: drivers: rtc: rtc_api: support STM32 boards`:
Does it allows to cover all boards used on the bench ?

ajarmouni-st commented:
> Does it allows to cover all boards used on the bench ?

No, a `.overlay` & a `.conf` should be added for each board for it to be tested. Also the RTC source clock must be configured in `.dts`of each board. Alternatively, a specific test case can be added for STM32 boards in `testcase.yaml` with a common overlay & `.conf`.

ajarmouni-st committed:
drivers: rtc_stm32: add support for alarms A & B

Add support for STM32 RTC Alarms A & B for all series except F1X.

Signed-off-by: Abderrahmane Jarmouni <abderrahmane.jarmouni-ext@st.com>

ajarmouni-st committed:
drivers: rtc: stm32: fix clk enable call

On STM32WBAX series, there is no bit in BCDR register to enable RTC.
Enabling RTC is done directly via the RCC APB register bit.

Signed-off-by: Abderrahmane Jarmouni <abderrahmane.jarmouni-ext@st.com>

erwango commented on boards/st/nucleo_h723zg/nucleo_h723zg.dts:
```diff
@@ -124,7 +126,7 @@
 
 &rtc {
 	clocks = <&rcc STM32_CLOCK_BUS_APB4 0x00010000>,
-		 <&rcc STM32_SRC_LSI RTC_SEL(2)>;
+		 <&rcc STM32_SRC_LSE RTC_SEL(1)>;
```
Either split in a dedicated commit or add a line in commit message to explain that you change clock for xxx reason.
Not that splitting is preferred in general as it allows to easily revert in case of trouble.

erwango reviewed (changes requested):
We're getting closer.

Aside from the comments, in commit "tests: drivers: rtc: rtc_api: support STM32 boards", it seems that some of the bench boards are still missing

ajarmouni-st committed:
drivers: rtc: stm32: make it clean again

Make driver clean again.

Signed-off-by: Abderrahmane Jarmouni <abderrahmane.jarmouni-ext@st.com>

ajarmouni-st committed:
drivers: rtc: stm32: subsecond not supported by F2x

subsecond counting is not supported by STM32F2x SoC series.

Signed-off-by: Abderrahmane Jarmouni <abderrahmane.jarmouni-ext@st.com>

ajarmouni-st committed:
boards: st: nucleo_f103rb: fix doc

STM32 RTC driver does not support F1x series for now.

Signed-off-by: Abderrahmane Jarmouni <abderrahmane.jarmouni-ext@st.com>

bjarki-andreasen reviewed (approved):
Nice work :)

721082 commented:
Great! Is `RTC_ALARM_TIME_MASK_NSEC` support when `HW_SUBSECOND_SUPPORT==1` planned in another step?

ajarmouni-st commented:
> Great! Is `RTC_ALARM_TIME_MASK_NSEC` support when `HW_SUBSECOND_SUPPORT==1` planned in another step?

Yes.
